File: src/cpp/step1.cpp
#include "printer.hpp"
#include "reader.hpp"

#include <iostream>
#include <string>

static auto READ(auto in) {
  return mal::read_str(in);
}
static auto EVAL(auto in) {
  return in;
}
static auto PRINT(auto in) {
  return mal::pr_str(in);
}

static auto rep(auto in) {
  return PRINT(EVAL(READ(in)));
}

static auto readline(std::string & line) {
  std::cout << "user> ";

  return static_cast<bool>(std::getline(std::cin, line));
}
int main() {
  std::string line;
  while (readline(line)) {
    std::cout << rep(line) << "\n";
  }
}

File: src/ocaml/step1.ml
let read str : string = str
let eval str : string = str
let print str : string = str

let rep str : string = 
  str |> read |> eval |> print
;;

try
  while true do
    print_string Reader.prompt;
    read_line () |> rep |> print_string;
    print_newline ();
  done;
with
  End_of_file -> ()
